src/modipulate-gml/modipulate_gml.c

```diff
@@ -0,0 +1,372 @@
+/* modipulate_gml.c
+ *
+ * Copyright 2015 Eric Gregory and Stevie Hryciw
+ */
+
+/* ------------------------------------------------------------------------ */
+
+#include <stdio.h>
+#include "modipulate.h"
+#include "modipulate_gml.h"
+
+/* ------------------------------------------------------------------------ */
+
+#define NOSONG ((ModipulateSong)0)
+
+#define MAX_SONGS 100
+
+typedef enum {
+    ERR_OK = 0,
+    LOADED_OK = -1000,
+
+    ERR_MIN = -100,
+    ERR_ASSERT,
+    ERR_FAIL,
+    ERR_INVALIDSONGID,
+    ERR_SONGOUTOFRANGE,
+    ERR_CHANNELOUTOFRANGE,
+    ERR_AMPOUTOFRANGE,
+    ERR_VALUEOUTOFRANGE,
+    ERR_FULL,
+} MgmlError;
+
+#define CMD_VOL 0
+#define CMD_FX  1
+
+/* ------------------------------------------------------------------------ */
+
+static char errbuf[1024] = "";
+static ModipulateSong songs[MAX_SONGS] = {0};
+
+/* Set song pointer. Return error if applicable.
+ */
+static int get_song(double songid, ModipulateSong* song) {
+    unsigned int id;
+
+    if (songid < 0.0 || songid >= MAX_SONGS) {
+        return ERR_SONGOUTOFRANGE;
+    }
+    id = songid;
+    if (songs[id] == NOSONG) {
+        return ERR_INVALIDSONGID;
+    }
+    *song = songs[id];
+
+    return ERR_OK;
+}
+
+/* ---- Global ------------------------------------------------------------ */
+
+double modipulategml_global_init(void) {
+    ModipulateErr err = modipulate_global_init();
+
+    return (err == MODIPULATE_ERROR_NONE ? LOADED_OK : ERR_FAIL);
+}
+
+double modipulategml_global_deinit(void) {
+    ModipulateErr err = modipulate_global_deinit();
+
+    return (err == MODIPULATE_ERROR_NONE ? ERR_OK : ERR_FAIL);
+}
+
+char* modipulategml_error_to_string(double errno) {
+    char* moderr = modipulate_global_get_last_error_string();
+
+    switch ((int)errno) {
+        case ERR_OK:
+        case LOADED_OK:
+            return "Everything is OK";
+        case ERR_ASSERT:
+            return "Error in modipulate-gml code";
+        case ERR_FAIL:
+            snprintf(errbuf, sizeof (errbuf),
+                "Internal Modipulate error: %s", moderr ? moderr : "?");
+            return errbuf;
+        case ERR_INVALIDSONGID:
+            return "Song ID does not point to a loaded song";
+        case ERR_SONGOUTOFRANGE:
+            return "Song ID is outside of valid range";
+        case ERR_CHANNELOUTOFRANGE:
+            return "Song channel is outside of valid range";
+        case ERR_AMPOUTOFRANGE:
+            return "Amplitude is outside of valid range";
+        case ERR_VALUEOUTOFRANGE:
+            return "Supplied value is outside of valid range";
+        case ERR_FULL:
+            return "No more capacity to load songs";
+        default:
+            return "Unknown error";
```

Comment:
@MrEricSir Do you mean for the default case? The library never gives out an unknown error code, so a debug assert would be ok, but obviously the application can call this with any number it wants so it shouldn't abort in release.